---
title: "Normalize sequencing data thanks to Pareto distribution"
output: 
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Installation

First of all we need to install PsiNorm:

```{r, eval = FALSE}
if(!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("PsiNorm")
```

```{r}
suppressPackageStartupMessages(
  {
    library(SingleCellExperiment)
    library(splatter)
    library(SummarizedExperiment)
    library(S4Vectors)
    library(BiocGenerics)
    library(ggplot2)
    library(MatrixGenerics)
    library(scater)
    library(Rtsne)
    library(cluster)}
)
```

## Introduction

PsiNorm is a new scalable between-normalization for single cell RNA-seq count data based on the power-law Pareto type I  distribution parameter estimate. It can be demonstrated that the Pareto parameter is inversely proportional to the sequencing depth, it is sample specific and its estimate can be obtained for each cell independently. PsiNorm computes the shape parameter for each cellular sample and then uses it as multiplicative size factor to normalize the data. The final goal of the transformation is to align the gene expression distribution especially for those genes characterised by high expression. Note that, similar to other global scaling methods, our method does not remove batch effects, which can be dealt with downstream tools. 

To evaluate the ability of PsiNorm to remove technical bias and  reveal the true cell similarity structure, we used both an unsupervised and a supervised approach.
We first simulate a scRNA-seq experiment with two known clusters using $spaltter$ bioC package.  Then in the unsupervised approach, we i) reduce dimentionality using tSNE, ii) identify clusters using $clara$ partitional method and then we iii)  computed the Adjusted Rand Index (ARI) to compare the known and the estimated partition.

In the supervised approach, we i) reduce dimentionality using tSNE, and we  ii)  computed the silhouette index of the  known partition in the reduced dimensional space. 

## Data Simulation

We simulate a matrix of counts with 2000 cellular samples amd 10000 genes with splatter BioC package:

```{r}
set.seed(1234)
params <- newSplatParams()
N=2000
sce <- splatSimulateGroups(params,batchCells=N,
                           group.prob = rep(0.2,5),
                           de.prob = 0.1,
                           verbose = FALSE) 
```

sce is a SingleCellExperiment object with a single batch and five different cellular groups.

To visualize the data we used the first two TSNE dimensions estimated starting from the raw count matrix.  
```{r}
set.seed(1234)
sce<-runTSNE(sce, exprs_values="counts", scale=T)
df<-data.frame(TSNE1=reducedDim(sce)[,1],
               TSNE2=reducedDim(sce)[,2],
               group=sce$Group)
ggplot(df,aes(TSNE1, TSNE2, color=group))+
  geom_point()+
  theme_classic()+
  theme(legend.title = element_blank())
```

From the plot it is clear that known groups can not be recognised using raw counts. 


## Data Normalization with PsiNorm 

To normalize the raw count we used  PsiNorm normalization (pareto_norm function) and we visualized the data using the first two tSNE components. 

```{r}
sce<-PsiNorm::pareto_norm(sce) #Â NON SAREBBE MEGLIO CHIAMARE LA FUNZIONE COME IL NOME DELLA NORMALIZZAIZONE ? PsiNorm ? 
set.seed(1234)
sce<-runTSNE(sce, exprs_values="Pareto", scale=T)
df<-data.frame(TSNE1=reducedDim(sce)[,1],
               TSNE2=reducedDim(sce)[,2],
               group=sce$Group)
ggplot(df,aes(TSNE1, TSNE2, color=group))+
  geom_point()+
  theme_classic()+
  theme(legend.title = element_blank())
```

We can appreciate from the plot that  PsiNorm allows a  better separation among known cellular groups.


## Unsupervised approach: ARI index 

We calculate ARI index of both raw counts and PsiNorm normalized counts after tSNE dimension reduction and $clara$ clustering (with $k$ equal to the simulated number of clusters); higher the Silhouette, better the normalization. 

DA FARE 


## Supervised approach: Silhouette index

We calculate Silhouette index of both raw counts and PsiNorm normalized counts after tSNE dimension reduction exploiting known simulated clusters; higher the Silhouette, better the normalization. 

```{r}
set.seed(1234)
sce<-runTSNE(sce, exprs_values="counts", scale=T)
dist<-daisy(reducedDim(sce, "TSNE"))
dist<-as.matrix(dist)
print(paste("Silhouette from raw counts:", round(summary(
    silhouette(x=as.numeric(as.factor(sce$Group)),
               dmatrix = dist))$avg.width, digits = 3)))
sce<-runTSNE(sce, exprs_values="Pareto", scale=T)
dist<-daisy(reducedDim(sce, "TSNE"))
dist<-as.matrix(dist)
print(paste("Silhouette from Pareto normalized counts:", round(summary(
    silhouette(x=as.numeric(as.factor(sce$Group)),
               dmatrix = dist))$avg.width, digits = 3)))
```
Pareto normalization considerably increses the Silhouette index. 


## Correlation of PC1 and PC2 with sequencing depth

To check if PsiNorm is able to capture technical noise and remove unwanted variation within a dataset (due for instance to differences in sequencing depth), we check whether the first two PCs are capturing technical variance. We computed the maximum correlation obtained between PC1 and PC2 and cell sequencing depths; a higher correlation indicates that the normalization was not able to properly remove noise.

```{r}
set.seed(4444)
PCA<-reducedDim(runPCA(sce, exprs_values="counts", scale=T), "PCA") 
PCAp<-reducedDim(runPCA(sce, exprs_values="Pareto", scale=T), "PCA")
depth<-apply(counts(sce), 2, sum)

print(paste("The Correlation from the raw data is:",
            round(abs(max(cor(PCA[,1], depth), cor(PCA[,2], depth))), digits=3)))
print(paste("The Correlation from the Pareto normalized data is:",
            round(abs(max(cor(PCAp[,1], depth), cor(PCAp[,2], depth))), digits = 3)))
```
Our results demonstrate that the correlation significantly decreases  after the PsiNorm normalization.

## Session Information
```{r}
sessionInfo()
```






